### PR TITLE
Fix compilation with latest `rustc` and lifetime issues

### PR DESCRIPTION
**Pull request by drbo**
This PR includes several fixes:
* The code did not compile with the beta version of `rustc`. To fix this, I used the crates.io `libc` library. This mandated changing to Cargo as the build system.
* Strings were not properly null-terminated and not properly converted to the modified UTF-8 encoding used by the JVM.
* It was possible to use a JVM that had been destroyed, resulting in undefined behavior.

Also, it adds some documentation to various functions and structs.

There are still more issues to fix:
* it is possible to pass the wrong type to a native method
* no method of generating JNI functions
* more documentation to write

I intend to fix these at a later date.

I added myself to the `authors` field in the last commit, due to having rewritten a large part of the code and my plans to contribute even more code. Feel free to not include this commit.

### CODE REVIEW

**Comment by Monnoroch** [comment]
I actually planned to return to this project after 1.0, but nice job, thanks!

File: lib/j_chars.rs [hunk 1]
```diff
@@ -0,0 +1,139 @@
...
+		&(self.contents)
+	}
+	pub fn as_ptr(&self) -> *const ::libc::c_char {
+		self.contents.as_ptr() as *const ::libc::c_char
+	}
+	pub unsafe fn from_raw_vec (data: Vec<u8>) -> Self {
+		JavaChars { contents: data, }
+	}
+}
+// vim: set noexpandtab:
```
Comment (Monnoroch) [inline lib/j_chars.rs]: Could you maby remove those? I don't feel good putting editor configs in the code.

File: lib/native.rs [hunk 2]
```diff
@@ -329,12 +328,12 @@ pub struct JNINativeInterface {
 	pub NewFloatArray:	   extern "C" fn(env: *mut JNIEnvImpl, len: jsize) -> jfloatArray,
 	pub NewDoubleArray:	  extern "C" fn(env: *mut JNIEnvImpl, len: jsize) -> jdoubleArray,
 
-	pub GetBooleanArrayElements: extern "C" fn(env: *mut JNIEnvImpl, array: jbooleanArray, isCopy: *mut jboolean) -> *mut jboolean,
-	pub GetByteArrayElements:	extern "C" fn(env: *mut JNIEnvImpl, array: jbyteArray, isCopy: *mut jboolean) -> *mut jbyte,
-	pub GetCharArrayElements:	extern "C" fn(env: *mut JNIEnvImpl, array: jcharArray, isCopy: *mut jboolean) -> *mut jchar,
-	pub GetShortArrayElements:   extern "C" fn(env: *mut JNIEnvImpl, array: jshortArray, isCopy: *mut jboolean) -> *mut jshort,
-	pub GetIntArrayElements:	 extern "C" fn(env: *mut JNIEnvImpl, array: jintArray, isCopy: *mut jboolean) -> *mut jint,
-	pub GetLongArrayElements:	extern "C" fn(env: *mut JNIEnvImpl, array: jlongArray, isCopy: *mut jboolean) -> *mut jlong,
+	pub GetBooleanArrayElements:	extern "C" fn(env: *mut JNIEnvImpl, array: jbooleanArray,	isCopy: *mut jboolean) -> *mut jboolean,
```
Comment (Monnoroch) [inline lib/native.rs]: As far as I conserned, tabs are a bad fit for alignment.

File: lib/j_chars.rs [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+				_ => unreachable!()
+			}
+		}
+		vector.push(b'\0');
+		JavaChars { contents: vector, }
+	}
+	pub fn to_string(&self) -> Option<String> {
+		let mut a = Vec::<u8>::with_capacity(self.contents.len());
+		let mut counter:usize = 0;
+		loop {
```
Comment (Monnoroch) [inline lib/j_chars.rs]: Are you sure, there are no standart decoders, or at least some library to do this?

**Comment by drbo** [comment]
I think I covered all of your concerns. There are still ways to invoke undefined behavior (pass wrong types to a JNI method), but this should be a big improvement.

I did find a library to convert from something _similar_ to what the JVM uses, but it does not handle NUL bytes the way the JVM does.

**Pull request merged**